File: II rok/I semestr/MIA/#Code 9/B. Delivery Bears.py
pobierz = lambda : tuple(map(int, input().split()))
def main():
    n, m, x = pobierz()
    weights = [[0 for _ in range(n)] for _ in range(n)]

    for i in range(m):
        ai, bi, ci = pobierz()
        ai -= 1
        bi -= 1
        weights[ai][bi] = ci

    visited = [False for _ in range(n)]
    temp = [[0 for _ in range(n)] for _ in range(n)]
    koniec = n-1
    flow = 0

    def dfs(a):
        if (a == koniec):
            return True
        if (visited[a]):
            return False

        visited[a] = True

        for i in range(n):
            if (temp[a][i] >= flow):
                if (dfs(i)):
                    temp[a][i] -= flow
                    temp[i][a] += flow
                    return True
        return False

    left = 0
    right = 1e6
    middle = 0
    for _ in range(100):
        middle = (left + right) / 2

        for ai in range(n):
            for bi in range(n):
                temp[ai][bi] = (weights[ai][bi] / middle)

        f = 1e9
        bear = 0
        while(f > 0):
            for ai in range(n):
                visited[ai] = False
            flow = f
            if (dfs(0)):
                bear += f
            else:
                f /= 2
        if (bear >= x):
            left = middle
        else:
            right = middle
    print(middle*x)


main()

File: II rok/I semestr/MIA/#Code 9/B. Delivery Bears.cpp
#include <iostream>
#include <iomanip>
using namespace std;

double temp[50][50];
double weights[50][50];
bool visited[50];
int n;
int flow;

bool dfs(int i) {
  if (i == n-1)
    return true;
  if (visited[i])
    return false;

  visited[i] = true;

  for (int j = 0; j < n; j++)
    if (temp[i][j] >= flow && dfs(j)) {
      temp[i][j] -= flow;
      temp[j][i] += flow;
      return true;
    }
  return false;
}


int main()
{
  int m, x;
  cin >> n >> m >> x;

  for (int i = 0; i < m; i++) {
    int ai, bi, ci;
    cin >> ai >> bi >> ci;
    ai--;
    bi--;
    weights[ai][bi] = ci;
  }

  double left = 0, right = 1e6;

  for (int i = 0; i < 100; i++) {
    double middle = (left + right) / 2;

    for (int ai = 0; ai < n; ai++)
      for (int bi = 0; bi < n; bi++)
        temp[ai][bi] = weights[ai][bi]/middle;

    long long bear = 0, f = 1e9;

    while (f) {
      for (int ai = 0; ai < n; ai++) {
        visited[ai] = false;
      }
      flow = f;
      if (dfs(0))
        bear += f;
      else
        f /= 2;
    }

    if (bear >= x)
      left = middle;
    else
      right = middle;
  }

  cout << setprecision(10) << left * x;
}
